--- apps/app_queue.c	(revisión: 185598)
+++ apps/app_queue.c	(revisión: 185599)
@@ -4404,6 +4404,11 @@
 						struct member tmpmem;
 						membername = NULL;
 
+						if (ast_strlen_zero(var->value)) {
+							ast_log(LOG_WARNING, "Empty queue member definition at line %d. Moving on!\n", var->lineno);
+							continue;
+						}
+
 						/* Add a new member */
 						ast_copy_string(parse, var->value, sizeof(parse));
 						
