--- apps/app_queue.c	2010-04-15 22:23:22.000000000 +0200
+++ apps/app_queue.c	2010-04-15 22:27:05.000000000 +0200
@@ -3992,7 +3992,7 @@
 				record_abandoned(&qe);
 				reason = QUEUE_TIMEOUT;
 				res = 0;
-				ast_queue_log(args.queuename, chan->uniqueid, "NONE", "EXITWITHTIMEOUT", "%d", qe.pos);
+				ast_queue_log(args.queuename, chan->uniqueid, "NONE", "EXITWITHTIMEOUT", "%d|%ld", qe.pos,(long) (time(NULL) - qe.start));
 				break;
 			}
 
@@ -4010,7 +4010,7 @@
 				record_abandoned(&qe);
 				reason = QUEUE_TIMEOUT;
 				res = 0;
-				ast_queue_log(args.queuename, chan->uniqueid, "NONE", "EXITWITHTIMEOUT", "%d", qe.pos);
+				ast_queue_log(args.queuename, chan->uniqueid, "NONE", "EXITWITHTIMEOUT", "%d|%ld", qe.pos,(long) (time(NULL) - qe.start));
 				break;
 			}
 			/* Make a periodic announcement, if enabled */
@@ -4023,7 +4023,7 @@
 				record_abandoned(&qe);
 				reason = QUEUE_TIMEOUT;
 				res = 0;
-				ast_queue_log(args.queuename, chan->uniqueid, "NONE", "EXITWITHTIMEOUT", "%d", qe.pos);
+				ast_queue_log(args.queuename, chan->uniqueid, "NONE", "EXITWITHTIMEOUT", "%d|%ld", qe.pos,(long) (time(NULL) - qe.start));
 				break;
 			}
 			/* Try calling all queue members for 'timeout' seconds */
@@ -4037,7 +4037,7 @@
 			if (noption && tries >= qe.parent->membercount) {
 				if (option_verbose > 2)
 					ast_verbose(VERBOSE_PREFIX_3 "Exiting on time-out cycle\n");
-				ast_queue_log(args.queuename, chan->uniqueid, "NONE", "EXITWITHTIMEOUT", "%d", qe.pos);
+				ast_queue_log(args.queuename, chan->uniqueid, "NONE", "EXITWITHTIMEOUT", "%d|%ld", qe.pos,(long) (time(NULL) - qe.start));
 				record_abandoned(&qe);
 				reason = QUEUE_TIMEOUT;
 				res = 0;
@@ -4067,7 +4067,7 @@
 				record_abandoned(&qe);
 				reason = QUEUE_TIMEOUT;
 				res = 0;
-				ast_queue_log(args.queuename, chan->uniqueid, "NONE", "EXITWITHTIMEOUT", "%d", qe.pos);
+				ast_queue_log(args.queuename, chan->uniqueid, "NONE", "EXITWITHTIMEOUT", "%d|%ld", qe.pos,(long) (time(NULL) - qe.start));
 				break;
 			}
 
