--- channels/chan_sip.c	(revision 46)
+++ channels/chan_sip.c	(working copy)
@@ -7543,6 +7543,13 @@
 		else
 			ast_build_string(&t, &maxbytes, "<dialog id=\"%s\">\n", p->exten);
 		ast_build_string(&t, &maxbytes, "<state>%s</state>\n", statestring);
+		
+		if((state & AST_EXTENSION_RINGING) && global_notifyringing && strstr(p->useragent,"THOMSON")){
+                        ast_build_string(&t, &maxbytes, "<local><identity display=\"%s\">%s</identity><target uri=\"%s\"/></local>\n", p->exten, mfrom, mfrom);
+                        ast_build_string(&t, &maxbytes, "<remote><identity display=\"%s\">sip:*8%s</identity><target uri=\"sip:*8%s\"/></remote>\n", "pickup", mto+4, mto+4);
+              }
+
+
 		if (state == AST_EXTENSION_ONHOLD) {
 			ast_build_string(&t, &maxbytes, "<local>\n<target uri=\"%s\">\n"
 			                                "<param pname=\"+sip.rendering\" pvalue=\"no\"/>\n"
