--- res/res_odbc.c	(revisión: 188148)
+++ res/res_odbc.c	(revisión: 188149)
@@ -252,9 +252,13 @@
 						ast_log(LOG_WARNING, "Limit should be a number, not a boolean: '%s'.  Setting limit to 1023 for ODBC class '%s'.\n", v->value, cat);
 						limit = 1023;
 					} else if (ast_false(v->value)) {
-						ast_log(LOG_WARNING, "Limit should be a number, not a boolean: '%s'.  Disabling ODBC class '%s'.\n", v->value, cat);
-						enabled = 0;
+						/* Limit=no probably means "no limit", which is the maximum */
+						ast_log(LOG_WARNING, "Limit should be a number, not a boolean: '%s'.  Setting limit to 1023 for ODBC class '%s'.\n", v->value, cat);
+						limit = 1023;
 						break;
+					} else if (limit > 1023) {
+						ast_log(LOG_WARNING, "Maximum limit in 1.4 is 1023.  Setting limit to 1023 for ODBC class '%s'.\n", cat);
+						limit = 1023;
 					}
 				} else if (!strcasecmp(v->name, "idlecheck")) {
 					sscanf(v->value, "%d", &idlecheck);
