--- include/asterisk/utils.h	(revisión: 199624)
+++ include/asterisk/utils.h	(revisión: 199856)
@@ -262,12 +262,12 @@
 		|| (sin1->sin_port != sin2->sin_port));
 }
 
-#define AST_STACKSIZE 240 * 1024
+#define AST_STACKSIZE (((sizeof(void *) * 8 * 8) - 16) * 1024)
 
 #if defined(LOW_MEMORY)
-#define AST_BACKGROUND_STACKSIZE 48 * 1024
+#define AST_BACKGROUND_STACKSIZE (((sizeof(void *) * 8 * 2) - 16) * 1024)
 #else
-#define AST_BACKGROUND_STACKSIZE 240 * 1024
+#define AST_BACKGROUND_STACKSIZE AST_STACKSIZE
 #endif
 
 void ast_register_thread(char *name);
